File: python/pov/pov.py
"""
Code to reorient a tree based on a given node. 
"""

from json import dumps


class Tree:
    """
    N-Ary tree class implementation with just string (label) as node values.
    """
    def __init__(self, label, children=None):
        """
        Class initializer
        """
        self.label = label
        self.children = children if children is not None else []

    def __dict__(self):
        """
        Return class as a dictionary
        """
        return {self.label: [child.__dict__() for child in sorted(self.children)]}

    def __str__(self, indent=None):
        """
        Return class as a string
        """
        return dumps(self.__dict__(), indent=indent)

    def __lt__(self, other):
        """
        Compare with another tree and check if one is less than the other based on label string compare
        """
        return self.label < other.label

    def __eq__(self, other):
        """
        Compare with another tree and check if they are equal (if the dictionaries are equal)
        """
        return self.__dict__() == other.__dict__()
    
    def __hash__(self) -> int:
        """
        Compute the hash code for this object
        """
        return hash(self.label, self.children)
    
    def find(self, root, value, path):
        """
        Find a value in the tree starting with a given root
        root = Label of node to start searching at.
        value = label we are looking for.
        path = list filled with path from root to value
        """
        if root.label == value:
            path.append(root)
            return True
        
        for child in root.children:
            ret = self.find(child, value, path)
            if ret:
                path.insert(0, root)
                return True
        return False

    def from_pov(self, from_node):
        """
        Rebuild the tree with the given node as root.
        from_node = The node that should be at root of the reoriented tree.
        """
        if self.label == from_node:
            return self
        
        path = []
        if not self.find(self, from_node, path):
            raise ValueError('Tree could not be reoriented')
        root = path[0]
        path.pop(0)
        while len(path) > 0:
            next_node = path.pop(0)

            root.children = [child for child in root.children if child.label != next_node.label]
            next_node.children.append(root)
            root = next_node
        return root

    def path_to(self, from_node, to_node):
        """
        Return the path from one node to the other. It goes from from_node up the tree
        potentially to the root and back down (if necessary) to to_node.
        - from_node = Label of node to start with
        - to_node = Label of node to end on.
        """
        if from_node == to_node:
            return [from_node]
        
        root_from = []
        if not self.find(self, from_node, root_from):
            raise ValueError('Tree could not be reoriented')

        root_to = []
        if not self.find(self, to_node, root_to):
            raise ValueError('No path found')
        
        from_path = [node.label for node in root_from]
        to_path = [node.label for node in root_to]
        
        current = ''
        switched = False
        index = len(from_path) - 1
        ret = []
        
        while not switched or index < len(to_path):
            if switched:
                current = to_path[index]
                index += 1
                ret.append(current)
            else:
                current = from_path[index]
                next_index = -1
                if current in to_path:
                    next_index = to_path.index(current)
                
                if next_index >= 0:
                    switched = True
                    index = next_index + 1
                else:
                    index -= 1


                # Turn around at root.
                if index < 0:
                    switched = True
                    index = 1           

                # Return the current item.
                ret.append(current)

        return ret

File: go/pov/pov.go
package pov

import "slices"

type Tree struct {
	// Add the needed fields here
	label    string
	children []*Tree
}

// New creates and returns a new Tree with the given root value and children.
func New(value string, children ...*Tree) *Tree {
	return &Tree{label: value, children: children}
}

// Value returns the value at the root of a tree.
func (tr *Tree) Value() string {
	return tr.label
}

// Children returns a slice containing the children of a tree.
// There is no need to sort the elements in the result slice,
// they can be in any order.
func (tr *Tree) Children() []*Tree {
	return tr.children
}

// String describes a tree in a compact S-expression format.
// This helps to make test outputs more readable.
// Feel free to adapt this method as you see fit.
func (tr *Tree) String() string {
	if tr == nil {
		return "nil"
	}
	result := tr.Value()
	if len(tr.Children()) == 0 {
		return result
	}
	for _, ch := range tr.Children() {
		result += " " + ch.String()
	}
	return "(" + result + ")"
}

// POV problem-specific functions

// Find a node in the tree
// @param root: Root of the tree to search from
// @param value: The label of the node we are searching for
// @param source: The path build so far.
// @returns: Path from root to node with the given value and boolean indicating success/failure
func find(root *Tree, value string, source []*Tree) (dest []*Tree, ok bool) {
	if root.label == value {
		return append(source, root), true
	}

	for _, child := range root.children {
		newPath, ok := find(child, value, source)
		if ok {
			return append([]*Tree{root}, newPath...), true
		}
	}
	return nil, false
}

// FromPov returns the pov from the node specified in the argument.
// @param from: The node that should be at root of the reoriented tree.
// @returns: The root of the reoriented tree.
func (tr *Tree) FromPov(from string) *Tree {
	if tr.label == from {
		return tr
	}
	path, ok := find(tr, from, []*Tree{})
	if !ok {
		return nil
	}
	root := path[0]
	path = slices.Delete(path, 0, 1)
	for len(path) > 0 {
		next_node := path[0]
		path = slices.Delete(path, 0, 1)
		for index, child := range root.children {
			if child.label == next_node.label {
				root.children = slices.Delete(root.children, index, index+1)
				break
			}
		}
		next_node.children = append(next_node.children, root)
		root = next_node
	}
	return root
}

// PathTo returns the shortest path between two nodes in the tree.
// @param from: Label of node to start with
// @param to: Label of node to end on.
// @returns: Slice of strings one element for each step from 'from' to 'to'.
func (tr *Tree) PathTo(from, to string) []string {
	if from == to {
		return []string{from}
	}

	root_from, ok := find(tr, from, []*Tree{})
	if !ok {
		return nil
	}

	root_to, ok := find(tr, to, []*Tree{})
	if !ok {
		return nil
	}

	from_path := make([]string, len(root_from))
	for index, node := range root_from {
		from_path[index] = node.Value()
	}

	to_path := make([]string, len(root_to))
	for index, node := range root_to {
		to_path[index] = node.Value()
	}

	current := ""
	switched := false
	index := len(from_path) - 1
	ret := []string{}

	for !switched || index < len(to_path) {
		if switched {
			ret = append(ret, to_path[index])
			index += 1
		} else {
			current = from_path[index]
			contains_index := slices.Index(to_path, current)
			if contains_index > -1 {
				index = contains_index + 1
				switched = true
			} else {
				index -= 1
			}

			// Turn around at root.
			if index < 0 {
				switched = true
				index = 1
			}

			// Return the current item.
			ret = append(ret, current)
		}
	}
	return ret
}
